lib/cocoapods/installer/analyzer.rb

```diff
@@ -249,48 +253,119 @@ def generate_target(target_definition, specs)
           end
         end
 
-        target.pod_targets = generate_pod_targets(target, specs)
-
+        target.pod_targets = pod_targets.select do |pod_target|
+          pod_target.target_definitions.include?(target_definition)
+        end
         target
       end
 
-      # Setup the pod targets for an aggregate target. Group specs and subspecs
-      # by their root to create a {PodTarget} for each spec.
+      # Setup the pod targets for an aggregate target. Deduplicates resulting
+      # targets by grouping by grouping by platform and subspec by their root
+      # to create a {PodTarget} for each spec.
       #
-      # @param  [AggregateTarget] target
-      #         the aggregate target
+      # @param  [Hash{Podfile::TargetDefinition => Array<Specification>}] specs_by_target
+      #         the resolved specifications grouped by target.
       #
-      # @param  [Array<Specification>] specs
-      #         the specifications that need to be installed.
+      # @return [Array<PodTarget>]
+      #
+      def generate_pod_targets(specs_by_target)
+        if config.deduplicate_targets?
```

Comment:
this should be factored into multiple methods. as it is, it's _incredibly_ complex